deep_qa/data/vocabulary.py

```diff
@@ -0,0 +1,111 @@
+from collections import defaultdict
+import codecs
+import logging
+
+import tqdm
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class Vocabulary:
+    """
+    A Vocabulary maps strings to integers, allowing for strings to be mapped to an
+    out-of-vocabulary token.
+
+    Vocabularies are fit to a particular dataset, which we use to decide which tokens are
+    in-vocabulary.
+
+    Vocabularies also allow for several different namespaces, so you can have separate indices for
+    'a' as a word, and 'a' as a character, for instance.  Most of the methods on this class allow
+    you to pass in a namespace; by default we use the 'tokens' namespace, and you can omit the
+    namespace argument everywhere and just use the default.
+    """
+    def __init__(self):
+        self._padding_token = "@@PADDING@@"
+        self._oov_token = "@@UNKOWN@@"
+        self._token_to_index = defaultdict(lambda: {self._padding_token: 0, self._oov_token: 1})
+        self._index_to_token = defaultdict(lambda: {0: self._padding_token, 1: self._oov_token})
+        self._finalized = False
+
+    def set_from_file(self, filename: str, oov_token: str="@@UNKNOWN@@", namespace: str="tokens"):
```

Comment:
There are a couple of considerations:

1. I am planning on changing the `fit_word_dictionary` method to use a counter object, removing the `increment_count` method.  Then this could be a `@classmethod` like `from_dataset()`, like you're suggesting.
2. The `set_from_file` only exists to support reading in a vocabulary from someone else's trained model (I added this recently when I was getting your language model stuff into DeepQA).  You're right that I could similarly make this a `@classmethod`, and I will.
3. You actually might want the vocabulary to be mutable after construction in some cases.  Roy is looking at modifying a loaded model, changing the embedding layer and possibly adding words to it, which would require changing a loaded vocabulary.  So I'm not as convinced about removing the methods that allow modifying the vocabulary (there's only one, right?).